File: benchmarks/csharp/BinaryTrees.cs
using Godot;

public partial class BinaryTrees : Benchmark
{
    // Based on https://github.com/hanabi1224/Programming-Language-Benchmarks/blob/main/bench/algorithm/binarytrees/1.cs
    class TreeNode
    {
        public TreeNode left;
        public TreeNode right;

        TreeNode(TreeNode left = null, TreeNode right = null)
        {
            this.left = left;
            this.right = right;
        }

        internal static TreeNode Create(int d)
        {
            return d == 0 ? new TreeNode()
                          : new TreeNode(Create(d - 1), Create(d - 1));
        }

        internal int Check()
        {
            int c = 1;
            if (right != null)
            {
                c += right.Check();
            }
            if (left != null)
            {
                c += left.Check();
            }
            return c;
        }
    }

    const int MinDepth = 4;
    public void CalculateBinaryTrees(int input)
    {
        int maxDepth = Mathf.Max(MinDepth + 2, input);

        int stretchDepth = maxDepth + 1;
        GD.Print($"stretch tree of depth {stretchDepth}\t check: {TreeNode.Create(stretchDepth).Check()}");

        TreeNode longLivedTree = TreeNode.Create(maxDepth);
        int maxPlusMinDepth = maxDepth + MinDepth;
        for (int depth = MinDepth; depth < maxDepth; depth += 2)
        {
            int iterations = 1 << (maxPlusMinDepth - depth);
            int check = 0;
            for (int i = 0; i < iterations; i++)
            {
                check += TreeNode.Create(depth).Check();
            }

            GD.Print($"{iterations}\t trees of depth {depth}\t check: {check}");
        }

        GD.Print($"long lived tree of depth {maxDepth}\t check: {longLivedTree.Check()}");
    }

    public void BenchmarkBinaryTrees13()
    {
        CalculateBinaryTrees(13);
    }

    public void BenchmarkBinaryTrees15()
    {
        CalculateBinaryTrees(15);
    }

    public void BenchmarkBinaryTrees18()
    {
        CalculateBinaryTrees(18);
    }
}

File: gdextension/src/benchmarks/binary_trees.cpp
#include "binary_trees.h"
#include <godot_cpp/core/class_db.hpp>

#include <godot_cpp/variant/utility_functions.hpp>

using namespace godot;

void CPPBenchmarkBinaryTrees::_bind_methods() {
    ClassDB::bind_method(D_METHOD("benchmark_binary_trees_13"), &CPPBenchmarkBinaryTrees::benchmark_binary_trees_13);
    ClassDB::bind_method(D_METHOD("benchmark_binary_trees_15"), &CPPBenchmarkBinaryTrees::benchmark_binary_trees_15);
    ClassDB::bind_method(D_METHOD("benchmark_binary_trees_18"), &CPPBenchmarkBinaryTrees::benchmark_binary_trees_18);
}

void CPPBenchmarkBinaryTrees::calculate_binary_trees(int input) {
    int max_depth = godot::Math::max(min_depth + 2, input);

    int stretch_depth = max_depth + 1;
    TreeNode* stretch_tree = TreeNode::create(stretch_depth);
    godot::UtilityFunctions::print("stretch tree of depth ", stretch_depth, "\t check: ", stretch_tree->check());
    delete stretch_tree;

    TreeNode* long_lived_tree = TreeNode::create(max_depth);
    int max_plus_min_depth = max_depth + min_depth;
    for (int depth = min_depth; depth < max_depth; depth += 2) {
        int iterations = 1 << (max_plus_min_depth - depth);
        int check = 0;
        for (int i = 0; i < iterations; i++) {
            TreeNode* check_tree = TreeNode::create(depth);
            check += check_tree->check();
            delete check_tree;
        }

        godot::UtilityFunctions::print(iterations, "\t trees of depth ", depth, "\t check: ", check);
    }

    godot::UtilityFunctions::print("long lived tree of depth ", max_depth, "\t check: ", long_lived_tree->check());
    delete long_lived_tree;
}

void CPPBenchmarkBinaryTrees::benchmark_binary_trees_13() {
	calculate_binary_trees(13);
}

void CPPBenchmarkBinaryTrees::benchmark_binary_trees_15() {
	calculate_binary_trees(15);
}

void CPPBenchmarkBinaryTrees::benchmark_binary_trees_18() {
	calculate_binary_trees(18);
}

CPPBenchmarkBinaryTrees::CPPBenchmarkBinaryTrees() {}

CPPBenchmarkBinaryTrees::~CPPBenchmarkBinaryTrees() {}
